clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3562,122 +3562,116 @@ static void emitCPPTypeString(raw_ostream &OS, QualType Ty) {
   emitWithoutAnonNamespaces(OS, Ty.getAsString(P));
 }
 
-static void printArguments(ASTContext &Ctx, raw_ostream &ArgOS,
-                           ArrayRef<TemplateArgument> Args,
-                           const PrintingPolicy &P);
-
-static void emitKernelNameType(QualType T, ASTContext &Ctx, raw_ostream &OS,
-                               const PrintingPolicy &TypePolicy);
-
-static void printArgument(ASTContext &Ctx, raw_ostream &ArgOS,
-                          TemplateArgument Arg, const PrintingPolicy &P) {
-  switch (Arg.getKind()) {
-  case TemplateArgument::ArgKind::Pack: {
-    printArguments(Ctx, ArgOS, Arg.getPackAsArray(), P);
-    break;
-  }
-  case TemplateArgument::ArgKind::Integral: {
-    QualType T = Arg.getIntegralType();
-    const EnumType *ET = T->getAs<EnumType>();
-
-    if (ET) {
-      const llvm::APSInt &Val = Arg.getAsIntegral();
-      ArgOS << "static_cast<"
-            << ET->getDecl()->getQualifiedNameAsString(
-                   /*WithGlobalNsPrefix*/ true)
-            << ">"
-            << "(" << Val << ")";
-    } else {
-      Arg.print(P, ArgOS);
+class SYCLKernelNameTypePrinter
+    : public TypeVisitor<SYCLKernelNameTypePrinter>,
+      public ConstTemplateArgumentVisitor<SYCLKernelNameTypePrinter> {
+  using InnerTypeVisitor = TypeVisitor<SYCLKernelNameTypePrinter>;
+  using InnerTemplArgVisitor =
+      ConstTemplateArgumentVisitor<SYCLKernelNameTypePrinter>;
+  raw_ostream &O;
+  PrintingPolicy &P;
+
+  void printTemplateArgs(ArrayRef<TemplateArgument> Args) {
+    for (size_t I = 0, E = Args.size(); I < E; ++I) {
```

Comment:
Just a suggestion/nit. It's probably not a concern for people familiar with clang code base or SemaSYCL, but I found myself scrolling up to see what "O", "P" , "E" were. If you find this reasonable you could follow the naming convention mentioned here: [llvm-naming-convention](https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly), else please feel free to ignore this comment.

Reply:
Yes, it sometimes happen in clang codebase, and I'm pretty used to it so I end up leaving a lot of non-informative names, you are right about it. However I think it should be ok for loop indices at least. I will go through my patch and see what can be changed.

Reply:
"OS" is typically used for 'output stream'.  "E" and "I" typically mean "End iterator" and "iterator" (both are K&R things).

So O->OS would be a good improvement, I think I/E for this is a fairly typical use case. 

I don't have any good idea on "P", Since the best name seems taken up by the type name :)

Reply:
Done.